### PR TITLE
Exclude gifts from getOrderTotal when selected ONLY_PRODUCTS

### LINKED ISSUE #22615: Discount - Incorrect display on FO with free gift + percentage cart rule
#### Describe the bug

On FO, when a cart rule with a free gift + percentage discount is added to the cart, the discount displayed on the line of this cart rule is incorrect, because the percentage discount is calculated also on gifted product price.
The Total discount is OK, and on BO all is OK too.

Might be related to #20690 and to #16673.

It was OK in 1.7.6.9.

#### Expected behavior

See step 5.

#### Steps to Reproduce

1. In BO > Catalog > Discounts, add a new cart rule
2. In the "Actions" part, set a percentage discount (15%) and a free gift (in my case, a mug at 100€ tax included). Save the cart rule.
3. Go to FO, add to the cart another product (at 100€ tax included too in my case)
4. In the cart, add the discount
5. Check the discount line : Discount is wrong (130€ instead of 115€), the percentage part of the discount is calculated on both paid AND gifted product (2x15€). It should be calculated only on the paid product (1x15€).
Note that the total discount is OK, so the total tax included of the order is also OK.

**Screenshots**

Cart rules Actions: 
bug discount percentage and gift cart rule

Display KO in FO: 130€ instead of 115€.
bug discount percentage and gift FO

OK in BO, on order and in invoice:
bug discount percentage and gift invoice
bug discount percentage and gift BO order

#### Additional information

* PrestaShop version: 1.7.7.0, 1.7.7.1 build 1
* PHP version: 7.2

### PR DESCRIPTION
**Pull request by PierreRambaud**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | 1.7.7.x
| Description?      | Exclude gift when selecting ONLY_PRODUCTS.
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #22615
| How to test?      | Follow ticket instruction.
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/22768)

### CODE REVIEW

**Review by jolelievre — DISMISSED** [review]

**Review by matthieu-rolland — DISMISSED** [review]

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

I have two issues:

First issue following these steps:

1. In BO > Catalog > Discounts, add a new cart rule
2. In the "Actions" part, set a percentage discount (15%) and a free gift (in my case, a mug at 100€ tax included). Save the cart rule.
3. Go to FO, add to the cart another product (at 100€ tax included too in my case)
4. In the cart, add the discount => The discount line => ok
5. Click on Proceed to checkout 
image
6. See the error 
image

The second issue following these steps:
1. In BO > Catalog > Discounts, add a new cart rule
2. In the "Actions" part, set a percentage discount (15%) and a free gift (in my case, a mug at 100€ tax included). Save the cart rule.
3. Go to FO, add to the cart the same gifted product
4. In the cart, add the discount => The discount line => NOK 
image
5. Click on Proceed to checkout 
image

Also, this [issue](https://github.com/PrestaShop/PrestaShop/issues/22615#issuecomment-755230771) is not fixed with this PR, it should be?
image

Thank you!

**Comment by PierreRambaud** [comment]
I fix the first one, didn't check more than the original issue :sweat_smile: 

> Also, this [issue](https://github.com/PrestaShop/PrestaShop/issues/22615#issuecomment-755230771) is not fixed with this

You're looking at the Order page, I talked about the Shopping Cart page :sweat_smile: 


> The second issue following these steps:

It's not the same issue and can be fixed in a new one. It occurs with or without my fix.

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

Thanks!
I moved the fixed issue to be reviewed in PrestaShop 1.7.7.x :slightly_smiling_face: 

Thank you!

**Comment by PierreRambaud** [comment]
> Hi @PierreRambaud,
> 
> Thanks!
> I moved the fixed issue to be reviewed in PrestaShop 1.7.7.x 
> 
> Thank you!

The pull request has already been reviewed ^^

**Comment by khouloudbelguith** [comment]
Hi @PierreRambaud,

It is ok :heavy_check_mark: 
Shopping cart page 
image
Checkout Page
image

About Other issues: 
https://github.com/PrestaShop/PrestaShop/issues/22946
https://github.com/PrestaShop/PrestaShop/issues/22948

Thank you!

File: classes/Cart.php [hunk 1]
```diff
@@ -2113,13 +2113,21 @@ public function getOrderTotal(
...
             foreach ($products as $key => $product) {
-                if ($product['is_virtual']) {
+                if (!empty($product['is_virtual'])) {
                     unset($products[$key]);
                 }
             }
             $type = Cart::ONLY_PRODUCTS;
         }
 
+        if ($type == Cart::ONLY_PRODUCTS) {
```
Comment (kpodemski) [inline classes/Cart.php:2123]: ha, it's interesting that it's working for you, sometimes we need to go through `CartRules` and check if `reduction_product` is something from cart products, but yeah we all know how it's going with cart rules 😅

Comment (PierreRambaud) [reply classes/Cart.php:2123]: The whole Cart & Cart rules have to be refactored :sweat:

**Comment by khouloudbelguith** [comment]
Hi @kpodemski,

Yes, you are right!
Thank you :slightly_smiling_face: 
Steps to reproduce the issue:
1. Create a product with price = 120Euro & specific price =20Euro tax included
image
2. In the "Actions" part, set a percentage discount (15%) and a free gift (in my case, a mug at 100€ tax included) and Enable the "Exclude discounted products
" option. Save the cart rule.
image
3. Go to FO > Add the last product created to the cart
4. Apply discount
5. See error
image
6. Continue checkout > See error
image

Ping @PierreRambaud what do you think?

Thanks!

**Comment by khouloudbelguith** [comment]
Hi @kpodemski,

As discussed with @PierreRambaud.
The recent issue is not related to this PR.
I manage to reproduce this issue: https://github.com/PrestaShop/PrestaShop/pull/22768#issuecomment-765324937 with PS1771.
I will create it in a new issue.

Thanks, @PierreRambaud, and @kpodemski.

**Review by Progi1984 — APPROVED** [review]

**Review by PululuK — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @PierreRambaud & @khouloudbelguith

**Pull request merged**